### PR TITLE
Update C# Pages in view of the new preferred inference API

### PR DESCRIPTION
**Pull request by yuslepukhin**
### Description
Re-work C# code samples with OrtValue API.
Bring `C# Tutorial: Basic` to index. 

The site is currently published [here](https://yuslepukhin.github.io/onnxruntime)

### Motivation and Context
Direct all future usage to `Ortvalue` API

### CODE REVIEW

**Comment by natke** [comment]
@yuslepukhin There is an extra `:` in the header of the tutorial. If you change C# tutorial: basic to Basic C# tutorial, you should be good on the navigation

File: docs/tutorials/csharp/basic_csharp.md [hunk 1]
```diff
@@ -1,36 +1,150 @@
...
+One can then create an OrtValue on top of the tensors data. 
+
+`ShapeUtils` class provides some help to deal with multi-dimensional indices for OrtValues.
+
+`OrtValue` based API provides direct native memory access in a type safe manner using `ReadOnlySpan<T>` and `Span<T>` stack bases structures.
+OrtValue is a universal container that can hold different ONNX types, such as tensors, maps, and sequences.
+It always existed in the onnxruntime library, but was not exposed in the C# API.
+
+As before, `OrtValues` can be created directly on top of the managed `unmanaged` (struct based blittable types)  arrays.
+Read MS documentation on `blittable` data types. onnxruntime C# API allows use of managed buffers for input or output.
```
Comment (natke) [inline docs/tutorials/csharp/basic_csharp.md:34]: Could add a link to https://learn.microsoft.com/en-us/dotnet/framework/interop/blittable-and-non-blittable-types

File: docs/tutorials/csharp/basic_csharp.md [hunk 2]
```diff
@@ -1,36 +1,150 @@
...
+It provides uniform access to data via `ReadOnlySpan<T>` and `Span<T>` structures, regardless of its location, managed or unmanaged.
+
+`DenseTensor` class can still be used for multi-dimensional access to the data since the new `Span` based API feature
+only a 1-D index. However, some reported a slow performance when using `DenseTensor` class multi-dimensional access.
+One can then create an OrtValue on top of the tensors data. 
+
+`ShapeUtils` class provides some help to deal with multi-dimensional indices for OrtValues.
+
+`OrtValue` based API provides direct native memory access in a type safe manner using `ReadOnlySpan<T>` and `Span<T>` stack bases structures.
+OrtValue is a universal container that can hold different ONNX types, such as tensors, maps, and sequences.
```
Comment (natke) [inline docs/tutorials/csharp/basic_csharp.md:30]: I would move this para up to the beginning.

File: docs/tutorials/csharp/basic_csharp.md [hunk 3]
```diff
@@ -1,36 +1,150 @@
...
 # C# Tutorial: Basic
 
-Here is simple tutorial for getting started with running inference on an existing ONNX model for a given input data. The model is typically trained using any of the well-known training frameworks and exported into the ONNX format. 
+Here is a simple tutorial for getting started with running inference on an existing ONNX model for a given input data.
+The model is typically trained using any of the well-known training frameworks and then exported into the ONNX format. 
+
+Note, that the following classed `NamedOnnxValue`, `DisposableNamedOnnxValue`, `FixedBufferOnnxValue` are going
+to be deprecated in the future. They are not recommended for new code.
+
+The new `OrtValue` based API is the recommended approach. The `OrtValue` API generates less garbage and is more performant.
```
Comment (natke) [inline docs/tutorials/csharp/basic_csharp.md:19]: Add an `OrtValue API` heading here

File: docs/tutorials/csharp/basic_csharp.md [hunk 4]
```diff
@@ -1,36 +1,150 @@
...
+Some scenarios indicated 4x performance improvement over the previous API and significantly less garbage.
+It provides uniform access to data via `ReadOnlySpan<T>` and `Span<T>` structures, regardless of its location, managed or unmanaged.
```
Comment (natke) [inline docs/tutorials/csharp/basic_csharp.md:21]: Then maybe a `Data Types` heading

File: docs/tutorials/csharp/basic_csharp.md [hunk 5]
```diff
@@ -1,36 +1,150 @@
...
+`ShapeUtils` class provides some help to deal with multi-dimensional indices for OrtValues.
+
+`OrtValue` based API provides direct native memory access in a type safe manner using `ReadOnlySpan<T>` and `Span<T>` stack bases structures.
+OrtValue is a universal container that can hold different ONNX types, such as tensors, maps, and sequences.
+It always existed in the onnxruntime library, but was not exposed in the C# API.
+
+As before, `OrtValues` can be created directly on top of the managed `unmanaged` (struct based blittable types)  arrays.
+Read MS documentation on `blittable` data types. onnxruntime C# API allows use of managed buffers for input or output.
+
+If output shapes are known, one can pre-allocate `OrtValue` on top of the managed or unmanaged allocations and supply
```
Comment (natke) [inline docs/tutorials/csharp/basic_csharp.md:36]: `Data shapes` heading?

File: docs/tutorials/csharp/basic_csharp.md [hunk 6]
```diff
@@ -1,36 +1,150 @@
...
+those OrtValues to be used as outputs. Due to this fact, the need for `IOBinding` is greatly diminished.
+
+String data is represented as UTF-16 string objects in C#. It will still need to be copied and converted to UTF-8 to the native
+memory. However, that conversion is now more optimized and is done in a single pass without intermediate byte arrays.
+The same applies to string `OrtValue` tensors returned as outputs. Character based API now operates on `Span<char>`,
+`ReadOnlySpan<char>`, and `ReadOnlyMemory<char>` objects. This adds flexibility to the API and allows to avoid unnecessary copies.
+
+Except some of the above deprecated API classes, nearly all of C# API classes are `IDisposable`.
```
Comment (natke) [inline docs/tutorials/csharp/basic_csharp.md:44]: `LIfe cycle` heading?

File: docs/tutorials/csharp/basic_csharp.md [hunk 7]
```diff
@@ -1,36 +1,150 @@
...
+GC can not operate on native memory or any other native resources.
+
+The `using` statement or a block is a convenient way to ensure that the objects are disposed.
+`InferenceSession` can be a long lived object and a member of another class. It eventually must also need to be disposed.
+This means, the containing class also would have to be made disposable to achieve this.
+
+OrtValue API also provides visitor like API to walk ONNX maps and sequences.
+This is a more efficient way to access Onnxruntime data.
 
 To start scoring using the model, open a session using the `InferenceSession` class, passing in the file path to the model as a parameter.
```
Comment (natke) [inline docs/tutorials/csharp/basic_csharp.md:70]: `Run a model with C#` heading?

File: docs/tutorials/csharp/basic_csharp.md (hunk 5 above)
Comment (yuslepukhin) [reply docs/tutorials/csharp/basic_csharp.md:36]: I would say: `Supplying OrtValues for output`

File: docs/tutorials/csharp/basic_csharp.md (hunk 7 above)
Comment (yuslepukhin) [reply docs/tutorials/csharp/basic_csharp.md:70]: Instead of `open a session`, say `instantiate InferenceSession class`

**Review by natke — APPROVED** [review]

**Pull request merged**